### PR TITLE
[Diffusion] No-D2H component offload via a layerwise-style host store

### PR DESCRIPTION
**Pull request by kevin-mii**
## Motivation

Component offload (`--component-residency <c>=component-offload`, and the legacy
`--*-cpu-offload` flags) swapped a component out with
`module.to("cpu", non_blocking=True)` — a full device-to-host copy of **immutable**
weights into fresh host memory on every single swap-out. Three problems this PR removes:

1. **A wasted D2H copy per swap-out.** Inference never mutates weights. Measured on
   MiniMax-H3's 9.7 GiB video VAE on an RTX 5090: 4.85 s on the first cycle, 0.37 s per
   cycle after; now ~0 (one device sync + a pointer rebind).
2. **Unbudgeted, kernel-locked host memory.** `module.to("cpu", non_blocking=True)`
   allocates a *pinned* destination, entirely behind `HostPinBudget`'s back — 9.7 GiB per
   offloaded VAE that the kernel can never reclaim, RssAnon never shows, and no
   accounting ever sees. Pinned offload memory now goes through the budget, and a denial
   degrades to the existing pageable storage without copying anything.
3. **`keep_checkpoint_mapped` weights were destroyed after one use.** A VAE loaded as
   mmap views into its safetensors file (so a constrained host keeps those bytes
   evictable) lost the mapping permanently on the first GPU round trip. The mapping now
   survives: a mapped weight's host copy *is* its mmap view, never replaced.

## Modifications

Five commits:

1. **BCG × offloaded-DiT guard** (`runtime/server_args/server_args.py`). Any explicitly
   non-resident DiT placement with `--enable-breakable-cuda-graph` fails at startup
   validation, next to the other BCG rules: BCG captures the DiT forward, every offload
   flavor rebinds or frees its weight storage between forwards, and a captured graph
   replays capture-time addresses. An auto-selected placement disables BCG with a
   warning; non-DiT components stay free to offload.
2. **Pin-budget divisor** (`runtime/managers/memory_managers/host_memory_budget.py`).
   `HostPinBudget` divides by the node-local worker count (`num_gpus // nnodes`) so N
   ranks stop booking the same host headroom N times; the reserve is taken from the
   per-rank share. A denial degrades to pageable hosting, while the overcommit it
   prevents ends in unreclaimable pinned memory and the OOM killer. Single-GPU
   deployments are unaffected.
3. **Dead-line removal** in `layerwise_offload._ensure_mapped_courier`.
4. **Regression test**: a component-offload GPU round trip must preserve the checkpoint
   mapping (fails on the old strategy, flipped by the next commit).
5. **The host-store rewrite**
   (`runtime/managers/memory_managers/component_residency_strategies.py`). Adopts
   layerwise offload's own storage model at component granularity: a
   `ComponentHostStore` captures the checkpoint-time host weights once (mmap views stay
   mmap views; the rest is pinned within a `HostPinBudget` grant, falling back to
   pageable when the budget declines) and hangs off the module. Swap-in copies
   host-to-device on the prefetch stream and rebinds via `.data`; swap-out rebinds every
   parameter to a shared one-element device placeholder — exactly a released layerwise
   layer, so no D2H copy and no new host allocation, ever. The store survives
   strategy-cache rebuilds, releases its device copy on sleep, and stays coherent with
   writers through the layerwise contracts: weight refits route through
   `update_host_weights` (the `update_cpu_weights` analog) and LoRA merge/unmerge wraps
   mutation in `begin`/`end_host_update`, which re-resolves targets after layer
   replacement. The residency report walks the store like `layerwise_offload_managers`.
   DTensor components work through the same `to_local_tensor`/`wrap_for_target` helpers
   layerwise uses. MPS and FSDP routing are unchanged.

## Accuracy Tests

No numerics change — the same weight bytes reach the device; only where they rest and how
they travel changes. End-to-end MiniMax-H3 (fl2va t2va, 672x384/4 s/8 steps, cookbook
5090 flags with `video_vae=component-offload`, 3 consecutive requests per side) produced
identical outputs on old and new sides.

## Speed Tests and Profiling

Component offload is a cookbook-recommended placement for encoder/VAE-sized components.
The real 9.7 GiB H3 video-VAE checkpoint (mmap-loaded, `assign=True`) driven through
`ComponentOffloadStrategy` prepare/finish cycles on a 1× RTX 5090 (32 GB):

| per swap cycle | old strategy | this PR |
|---|---|---|
| swap-out wall | 4.85 s first, 0.37 s after (full D2H) | **~0 s** (sync + rebind) |
| swap-in wall (steady) | 0.37 s (from its pinned copy) | 0.56 s (straight from the mapping) |
| host state at rest | 9.7 GiB **pinned**, unbudgeted; mapping destroyed | 9.7 GiB **still checkpoint-mapped**; zero committed host bytes |
| device bytes per swap-in | 9.71 GiB | 9.71 GiB |

The mapped host copy trades ~0.2 s of swap-in against holding zero committed host
memory; weights the budget grants pinning keep the pinned fast path on both directions.
End-to-end H3 latency is neutral to within noise — the change is memory correctness, not
a latency claim.

Notes:
- `text_encoder=component-offload` cannot work on a 32 GB card for H3 (66.7 GB bf16
  encoder; whole-module swap-in OOMs — verified empirically). The cookbook's layerwise
  placement for it is the right one.
- 5 tests under `test/unit/realtime/`, `test/unit/sana_wm/`, and
  `test_consistency_metrics.py` fail on origin/main, unrelated to this PR; the companion
  `diffusion-fix-unit-tests` branch fixes them separately.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests). (One case per failure mode: the keep-mapped round trip; residency buckets stable across swap cycles with values intact; budget denial degrades without copying; both writer paths reach the host store; the BCG guard's raise/disable/ignore cases; the N-rank budget divisor and per-rank-share reserve. Full `multimodal_gen/test/unit` on the 5090 is green except the 5 pre-existing main failures fixed in the companion branch — both together: 2191 passed, 0 failed.)
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations). (No flag semantics change; existing docs remain accurate.)
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed). (See the sections above.)
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #34220559150](https://github.com/sgl-project/sglang/actions/runs/34220559150)
Latest PR Test (Extra): :x: [Run #34220558891](https://github.com/sgl-project/sglang/actions/runs/34220558891)
Latest PR Test (AMD ROCm 7.2): :x: [Run #34220559171](https://github.com/sgl-project/sglang/actions/runs/34220559171)

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/managers/memory_managers/component_residency_strategies.py [hunk 1]
```diff
@@ -127,20 +121,283 @@ def prepare_for_use(
...
+        pin_budget: HostPinBudget,
+        component_name: str,
+    ) -> None:
+        regions = MappedRegions()
+        for name, target in chain(module.named_parameters(), module.named_buffers()):
+            local = to_local_tensor(target)
+            # a module captured warm pays one D2H here
+            host = (local if local.device.type == "cpu" else local.to("cpu")).detach()
+            self._targets[name] = target
+            self._host[name] = host
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/managers/memory_managers/component_residency_strategies.py:187]: Please retain a low-host-memory fallback for non-mmap weights. `_host` keeps the full CPU copy while `load()` allocates GPU weights; a denied `HostPinBudget` only disables pinning, not retention. This can increase runtime host usage and is particularly risky on shared-memory GPUs. Can we gate retention on host/shared-pool headroom and add a constrained-memory test?

File: python/sglang/multimodal_gen/runtime/managers/memory_managers/component_residency_strategies.py [hunk 2]
```diff
@@ -127,20 +121,283 @@ def prepare_for_use(
...
+                chosen_bytes += nbytes
+        if (
+            chosen
+            and torch.get_device_module().is_available()
+            and pin_budget.request(
+                component_name=component_name, weight_bytes=chosen_bytes
+            )
+        ):
+            for name in chosen:
+                self._host[name] = _pinned_like(self._host[name])
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/managers/memory_managers/component_residency_strategies.py:222]: Could we avoid accumulating both complete pageable and pinned copies during `_capture()`? `_targets` still reference the original CPU storages until `release()` runs after this loop, so pinning all eligible weights can temporarily hold roughly two host copies. Please release/rebind each eligible unshared source after copying, or explicitly account for and test this initialization peak.

**Review by mickqian — COMMENTED** [review]
Removing redundant D2H makes sense, but please report peak host/cgroup memory, pinned bytes and E2E for mapped and non-mapped weights under a real memory cap, including initialization and GPU execution. Also clarify "zero committed host bytes": mmap avoids a new anonymous/pinned copy, not physical RAM usage. Resident file-backed pages still consume memory and may be charged to the cgroup.
